core/src/main/scala/org/bykn/bosatsu/SourceConverter.scala

```diff
@@ -207,33 +225,38 @@ final class SourceConverter(
                 "flat_map_List"
             }
             val opExpr: Expr[Declaration] = Expr.Var(pn, Identifier.Name(opName), l)
-            val resExpr: Expr[Declaration] = (res, filter) match {
-              case (itOrSp, None) =>
-                apply(itOrSp.value)
-              case (itOrSp, Some(cond)) =>
-                val empty: Expr[Declaration] =
-                  Expr.Var(pn, Identifier.Constructor("EmptyList"), cond)
-                // we need theReturn
-                val r = itOrSp.value
-                val ritem = apply(r)
-                val sing = itOrSp match {
-                  case SpliceOrItem.Item(_) =>
-                    Expr.App(
-                      Expr.App(Expr.Var(pn, Identifier.Constructor("NonEmptyList"), r), ritem, r),
-                      empty,
-                      r)
-                  case SpliceOrItem.Splice(_) => ritem
-                }
+            val resExpr: Result[Expr[Declaration]] =
+              (res, filter) match {
+                case (itOrSp, None) =>
+                  apply(itOrSp.value)
+                case (itOrSp, Some(cond)) =>
+                  val empty: Expr[Declaration] =
+                    Expr.Var(pn, Identifier.Constructor("EmptyList"), cond)
+                  // we need theReturn
+                  val r = itOrSp.value
```

Comment:
`apply(itOrSp.value)` appears in both branches. Might be clearer to only match on filter (and a bit more efficient).